### PR TITLE
Add kubernetes Ingress backend

### LINKED ISSUE #11: Kubernetes backend
Any plans to add a configuration backend using Kubernetes annotations?

### PR DESCRIPTION
**Pull request by emilevauge**
This PR adds Kubernetes Ingress backend.

> This implementation is working, but there is a major issue, the outrageous dependencies needed by the official k8s client: https://github.com/kubernetes/kubernetes/issues/11606.
> 
> Unfortunately, this leads to a binary that weights **43M** (compared to **18M** initially)!
> ```
> -rwxr-xr-x 1 emile emile  43M févr.  9 19:50 traefik*
> ```
> 
> You are welcome to propose alternatives ;) I will not merge this one until a decent size is back.

As it seems that Kubernetes will not fix this issue soon, I decided to copy the API types into traefik and rewrite a Kubernetes Client. This allows traefik to keep its original size.  

- [x] initial implementation
- [x] clean code
- [x] find alternative to k8s official API
- [x] create an example k8s resources file
- [x] doc

Fixes #11 
Fixes #186

### CODE REVIEW

**Comment by philk** [comment]
You're not alone: https://github.com/kubernetes/kubernetes/issues/11606

**Comment by tboerger** [comment]
Maybe provide it with a build tag? Than you can build and release a slim version without k8s and a bloated version including k8s.

**Comment by vdemeester** [comment]
@tboerger I'm :+1: on that :angel:

**Comment by emilevauge** [comment]
Yeah, I was thinking about that too. This could be a temporary solution waiting for a lighter client.

**Comment by tboerger** [comment]
> Yeah, I was thinking about that too. This could be a temporary solution waiting for a lighter client.
>  Hide all checks

Sure, it can always be removed if the client shrinks sizewise.

File: provider/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+// Provide allows the provider to provide configurations to traefik
+// using the given configuration channel.
+func (provider *Kubernetes) Provide(configurationChan chan<- types.ConfigMessage) error {
+	//var KubernetesFuncMap = template.FuncMap{}
+	var ingClient client.IngressInterface
+	config := &client.Config{
+		Host: provider.Endpoint,
+	}
```
Comment (ekozan) [inline provider/kubernetes.go]: Maybe it's better in firstly case to try in cluster config 
`
config, err := client.InClusterConfig()
`

**Comment by Norbell** [comment]
A few days ago Kubernetes 1.2 was released. 
Is it possible that the newest Kubernetes client has less dependencies? 

Yesterday i tried to understand your code and found the following line.
What's the use of this "Limiter"?
`rateLimiter := util.NewTokenBucketRateLimiter(0.1, 1)`

By the way, the method "NewTokenBucketRateLimiter()" is no longer part of the "k8s.io/pkg/util/" package.
[https://godoc.org/github.com/kubernetes/kubernetes/pkg/util](url)

**Comment by emilevauge** [comment]
> Is it possible that the newest Kubernetes client has less dependencies?

Not really :'(
`-rwxrwxr-x 1 emile emile  49M mars  31 13:57 traefik*`

**Comment by jimmidyson** [comment]
I had this exact same issue when adding Kubernetes discovery to [Prometheus](http://prometheus.io). What I ended up doing is copying over the types I needed (a tiny subset of the whole Kubernetes API) & just using a standard `http.Client` with chunking for watches on service endpoints. This seems like it should work perfectly well here too.

**Comment by guybrush** [comment]
what about a server that runs parallel to traefik, which syncs kubernetes-ingress via the traefik-file (just a go-template)? the go-template could be provided via a kubenetes-configmap, so one can define or update the templates by using `kubectl {create,apply} traefik.yml`.

that way traefik would no need any k8s-dependencies (less binary-size AND loc) and traefik would not need to be updated everytime k8s changes ingress-behavior :D

**Comment by guybrush** [comment]
something like this:

``` yaml
# k8s-ingress-controller-traefik.yml

apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx-ingress-controller
  labels:
    k8s-app: ingress-controller-traefik
spec:
  replicas: 1
  selector:
    k8s-app: ingress-controller-traefik
  template:
    metadata:
      labels:
        k8s-app: ingress-controller-traefik
    spec:   
      containers:
      
      # traefik is running as load-balancer
      
      - image: traefik
        ports:
        - containerPort: 80
          hostPort: 80
        - containerPort: 443
          hostPort: 4444
        command: /traefik -c /var/traefik-file/traefik.yml
        volumeMounts:
        - name: data
          mountPath: /var/traefik-file
          
      
      # kube2traefik will sync k8s-ingress-resources to the file:
      # /var/traefik-file/traefik.yml
      # also kube2traefik subscribes to k8s-api for a configmap
      # with the name "ingress-controller-traefik-configmap"
      
      - image: kube2traefik
        volumeMounts:
        - name: data
          mountPath: /var/traefik-file
          
      volumes:
      - name: data
        emptyDir: {}
        
---

apiVersion: v1
kind: ConfigMap
metadata:
  name: ingress-controller-traefik-configmap
data:
  template: "<golang-template for traefik-config-file>"
```

**Comment by jimmidyson** [comment]
Yeah sounds good. I'd love to see native kubernetes support, but I see some other issues that will come up like config providers not being able to pass certificates for SNI.

**Comment by errm** [comment]
Romulus already has some kind of traefik support. https://github.com/albertrdixon/romulus

But it would be really nice to have batteries included.

**Comment by wernerb** [comment]
With regard to shrinking binaries, have you checked out these options? https://blog.filippo.io/shrink-your-go-binaries-with-this-one-weird-trick/

**Comment by emilevauge** [comment]
Hi all, I finally decided to copy the API ;)
This branch is now working with Kubernetes backend with a descent binary size:
```
-rwxrwxr-x 1 emile emile  16M avril 19 16:40 traefik*
```
I'm thinking of merging this one as is, but: for now, it doesn't support Ingress [TLS](http://kubernetes.io/docs/user-guide/ingress/#tls): as a workaround for the certs, we could use [secrets](http://kubernetes.io/docs/user-guide/secrets/) and mount the secrets as volumes in traefik container.

WDYT?

(PR description updated)

**Comment by rothgar** [comment]
I like the idea of using secrets for the certs. No need to implement something new or have the user learn anything other than what is already documented with kubernetes.

**Comment by jimmidyson** [comment]
Rather than mounting the secrets, which is only done on pod startup, I think you'll need to get the secret via the Kubernetes API, especially when ingresses are dynamically updated.

**Comment by emilevauge** [comment]
@jimmidyson totally agree :)

But, I'm not sure that the majority uses Kubernetes secrets to store their SSL certs.
Besides, you could use Let's Encrypt with Traefik to dynamically get your certs.

My proposal is to merge this one, as a first iteration.
I will add the ingress secrets support later.

Seems reasonable?

**Comment by mgoodness** [comment]
@emilevauge SGTM

**Comment by jimmidyson** [comment]
@emilevauge Absolutely.

**Comment by stongo** [comment]
@emilevauge +1 
I currently have Romulusd configured as an ingress controller and then use Let's Encrypt along with Traefik for SSL and it's totally great

**Comment by damoon** [comment]
i played a little bit the the example.
i found it would not work when:
a) namespaces get added or
b) the port changes (i used gcr.io/google_containers/echoserver:1.3 on port 8080)

i am looking forward to see traefik read its config directly from kubernetes.
great job :D

**Comment by emilevauge** [comment]
@damoon: thanks! Just pushed a fix that add namespace management and change rule `PathPrefix` -> `PathPrefixStrip`

**Comment by mholtzberg** [comment]
I just tested the implementation in worked right away, good work!
I wonder if it's possible to route directly to the pods instead of using the local kube-proxy as another layer in between. But if I understand correctly, that's how ingress is supposed to be used right now.
Is it already possible to register https letsencrypt support out of the ingress yaml?
Pretty nice so far!

**Comment by sstarcher** [comment]
As an example of k8s secrets for SSL this might help you out.  They have a similar setup, but it's k8s only.
https://github.com/deis/router#ssl

**Comment by mholtzberg** [comment]
@sstarcher: thanks, but that has nothing to do with traefik at all, or am I wrong? Actually I don't like to store TLS certs as secrets since pods need to restarted to get access to updated certs. I'd like to have traefik handle that.

File: configuration.go [hunk 2]
```diff
@@ -209,7 +210,11 @@ func LoadConfiguration() *GlobalConfiguration {
 	viper.AddConfigPath("$HOME/.traefik/") // call multiple times to add many search paths
 	viper.AddConfigPath(".")               // optionally look for config in the working directory
 	if err := viper.ReadInConfig(); err != nil {
-		fmtlog.Fatalf("Error reading file: %s", err)
+		if len(viper.ConfigFileUsed()) > 0 {
```
Comment (vdemeester) [inline configuration.go]: Why we don't do anymore `Fatalf` here ? :angel:

File: provider/k8s/client.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	}()
+	return watchCh, errCh, nil
+}
+
+// GetServices returns all services in the cluster
+func (c *clientImpl) GetServices(namespace string, predicate func(Service) bool) ([]Service, error) {
+	getURL := c.endpointURL + APIEndpoint + "/namespaces/" + namespace + "/services"
+
+	// Make request to Kubernetes API
+	request := gorequest.New().Get(getURL)
```
Comment (vdemeester) [inline provider/k8s/client.go]: There is some code duplication, but could be refactor afterward so, that's a nit :)

File: provider/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		token = string(tokenBytes)
+		log.Debugf("Kubernetes token: %s", token)
+	} else {
+		log.Debugf("Kubernetes load token error: %s", err)
+	}
+	caCert, err := ioutil.ReadFile(serviceAccountCACert)
+	if err == nil {
+		log.Debugf("Kubernetes CA cert: %s", serviceAccountCACert)
+	} else {
+		log.Debugf("Kubernetes load token error: %s", err)
```
Comment (vdemeester) [inline provider/kubernetes.go]: Only debug here ? If there is an error reading caCert, it's no big deal ? :sweat_smile:

**Comment by vdemeester** [comment]
Few comments / questios, but otherwise looks good :+1:

**Comment by ekozan** [comment]
@emilevauge for the k8s example can you add daemonset too. I think this kind more appropriate for this app

```yaml 
  apiVersion: extensions/v1beta1
  kind: DaemonSet
  name: traefik-ingress-controller
  labels:
    k8s-app: traefik-ingress-lb
  spec:
    template:
      metadata:
        labels:
          k8s-app: traefik-ingress-lb
      spec:
        nodeSelector: 
          app: traefik-ingress-lb
      terminationGracePeriodSeconds: 60
      containers:
      - image: containous/traefik
        name: traefik-ingress-lb
        imagePullPolicy: Always
        ports:
        - containerPort: 80
          hostPort: 80
        - containerPort: 443
          hostPort: 443
        args:
        - --kubernetes
        - --logLevel=DEBUG
```

File: configuration.go (hunk 2 above)
Comment (emilevauge) [inline configuration.go]: Because if we don't have any config file, it should continue, using evetually arguments ;)

Comment (vdemeester) [inline configuration.go]: ah, it's not related to k8s then :stuck_out_tongue_closed_eyes:

File: provider/k8s/client.go (hunk 3 above)
Comment (emilevauge) [inline provider/k8s/client.go]: fixed

File: provider/kubernetes.go (hunk 4 above)
Comment (emilevauge) [inline provider/kubernetes.go]: fixed

**Comment by emilevauge** [comment]
@ekozan: Why not, but your example is not working. [Official examples](https://github.com/kubernetes/contrib/tree/master/ingress/controllers) of ingress controllers are deployed as `ReplicationController`.
I'll keep the replication controller one for now ;)

**Comment by ekozan** [comment]
not only `ReplicationController`: https://github.com/kubernetes/contrib/blob/master/ingress/controllers/nginx/examples/daemonset/as-daemonset.yaml

hum ok forget it
I'll make a pull request after yours with a working example and readme

**Comment by stongo** [comment]
The only reason why DaemonSets should not be used is because currently there's no way to update or rolling-update them. Ideally as DaemonSets mature and allow updating, they would be the best choice as every host should really be running Traefik with dns round robin in front. But +1 to official example being an RC for now

**Comment by sstarcher** [comment]
Unless you are targeting K8S 1.1 I would recommend using a Replica Set over a Replica Controller in K8S 1.2

**Comment by vdemeester** [comment]
LGTM 🐻

**Pull request merged**